File: kappa-spring-boot-examples/src/main/java/com/github/erosb/kappa/examples/PetController.java
package com.github.erosb.kappa.examples;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.LocalDate;
import java.util.List;

record UserIdentifier(String id) {
}

record CreatePetRequest(String name, UserIdentifier owner, LocalDate birthDate) {
}

record User(int id, String firstName, String lastName) {
}

record Pet(int id, String name, User owner, long birthDate) {
}

@RestController
@RequestMapping("/api/pets")
public class PetController {

    @PostMapping
    void createPet(@RequestBody CreatePetRequest requestBody) {
        System.out.println("requestBody = " + requestBody);
    }

    @GetMapping
    List<Pet> getPets() {
        return List.of(
            new Pet(
                1,
                "",
                new User(2, "John", "Doe"),
                LocalDate.parse("2017-08-08").toEpochDay()
            )
        );
    }
}

File: src/main/kotlin/com/github/erosb/openapidemo/PetController.kt
package com.github.erosb.openapidemo

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

data class CreatePetRequest(var name: String)

@RestController
@RequestMapping("/api/pet")
class PetController {

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun createPet() {

    }
}
